sdk/component/runner.go

```diff
@@ -0,0 +1,202 @@
+package component
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os/signal"
+	"syscall"
+	"time"
+
+	_ "go.uber.org/automaxprocs"
+	"golang.org/x/sync/errgroup"
+)
+
+type (
+	// runner is used to setup the run context.
+	runner struct {
+		config *RunnerConfig
+	}
+
+	// componentRunnerFunc is an alias for executing component run functions.
+	componentRunnerFunc func(context.Context) error
+	// closerFunc is an alias for Close methods in components.
+	closerFunc func(ctx context.Context) error
+)
+
+// newRunner returns a new runner which is initialised looking at the default configuration
+// and can be customised applying options.
+func newRunner(opts ...RunnerOption) (*runner, error) {
+	cfg, err := newDefaultRunnerConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not create default runner configuration: %w", err)
+	}
+
+	r := &runner{config: cfg}
+
+	for _, opt := range opts {
+		if err := opt(r); err != nil {
+			return nil, err
+		}
+	}
+
+	return r, nil
+}
+
+// run is the core of the component SDK.
+// run performs all the steps needed to run components reliably and predictably.
+// run takes care of:
+// - signal termination: gracefully close components when a termination signal is detected.
+// - context cancellation: gracefully close components when a cancellation is detected.
+// - automatically setting GOMAXPROCS: Automatically set GOMAXPROCS to match Linux container CPU quota - https://github.com/uber-go/automaxprocs.
+// - panic handling.
+// - logging: initialisation and centralisation.
+// - running components: reliably and respecting cancellations.
+// - closing components: gracefully closing components to avoid inconsistencies.
+// - TODO: metrics.
+// - TODO: profiling.
+// - TODO: tracing.
+func run(
+	ctx context.Context,
+	componentRunner componentRunnerFunc,
+	closer closerFunc,
+	opts ...RunnerOption,
+) error {
+	r, err := newRunner(opts...)
+	if err != nil {
+		return fmt.Errorf("could not create runner: %w", err)
+	}
+
+	var (
+		conf   = r.config
+		logger = r.
+			config.Logging.Logger.
+			With(logKeySDKVersion, conf.SDKVersion).
+			With(logKeyComponentName, conf.ComponentName)
+		syncClose = make(chan struct{}, 1)
+		syncDone  = make(chan struct{}, 1)
+		syncErrs  = make(chan error, 1)
+	)
+
+	ctx, cancel := signal.NotifyContext(
+		ContextWithLogger(ctx, logger),
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT,
+		syscall.SIGABRT,
+		syscall.SIGKILL,
+	)
+	defer cancel()
+
+	g, ctx := errgroup.WithContext(ctx)
```

Comment:
this whole section looks like it's more complex than it's needed to be because of the use of an errgroup

I would expect that we would iterate over the stages of a workflow and for each stage a set of goroutines would be generated with each one sequentially executing the steps listed in a component, and all these goroutines would be added to a waitgroup. if any of the component steps throws an error the Close function is invoked to do whatever cleanup is needed.

using an errgroup means that we need to add all the channels for more synchronisation. instead we could just use a shared error with a mutex to protect a join function call, so that at the end we can have all the errors in one place.

components should be allowed to finish their job even if one of them fails, unless the user sends a kill signal to the process